[BREAKING] Drop constant and payable from ABI.

ekpyron opened the pull request:
Closes #4755.

Part of #3656.

To be discussed:
  - How to deal with breaking changes and the contracts in ``test/compilationTests/``?
  - Does it make sense to merge changes like this to the ``050`` branch at this point?

ekpyron committed:
Remove ``constant`` and ``payable`` from ABI.

 # tmp

ekpyron committed:
Update test suite.

ekpyron commented:
One of the reasons to create this was to see how the tests behave and start a discussion about that.
As expected, the tests involve external contracts, e.g. from truffle and the DAO, that are not yet forward compatible with changes like this, so we have to decide how to deal with this.

axic commented:
@ekpyron can we merge the `constant` -> `view` changes (docs, styleguide, std, test) separately on `develop`? That it not breaking.

ekpyron commented:
@axic That's true - I have them in separate commits here, so I'll extract them into a separate PR shortly!

ekpyron commented:
Once #4093 is merged to develop and develop is merged into 050, I'll rebase this PR.

axic commented on libsolidity/interface/ABI.cpp:
```diff
@@ -36,8 +36,6 @@ Json::Value ABI::generate(ContractDefinition const& _contractDef)
 		method["type"] = "function";
 		method["name"] = it.second->declaration().name();
-		// TODO: deprecate constant in a future release
-		method["constant"] = externalFunctionType->stateMutability() == StateMutability::Pure || it.second->stateMutability() == StateMutability::View;
```
We should put this warning into the documentation saying it will be removed with 0.5.0.

axic replied:
Done.

axic commented:
The ViewPureChecker must be updated to be turned on for `view` (there's a 0.5.0 check there).

ekpyron committed:
Adjust abi-spec doc.

ekpyron commented:
@axic The two last commits now enable enforcing ``view`` in ViewPureChecker. The last commit merely adjusts some end-to-end-tests and I extracted it to #4093 as well.
Actually I would have considered enabling these errors to be separate from this PR, but I guess it makes sense to add it in here.

ekpyron commented:
Of course, since this last change (``ViewPureChecker`` reports errors for ``view``), the tests will fail again and are much more involved to update. Since we decided to disable the compilation tests for 0.5.0 I will not update them any further.

ekpyron commented:
Rebased and removed the commit that updates ``test/compilationTests`` (since we decided to disable those tests for 0.5.0).

chriseth commented on docs/abi-spec.rst:
```diff
@@ -313,7 +313,10 @@ A function description is a JSON object with the fields:
+
+.. warning::
+    Prior to version 0.5.0 there was an additional field ``constant``. In 0.5.0 this field has been removed.
+    Instead you can check whether stateMutability is ``pure`` or ``view``.
```
Perhaps explain that `view` is the former `constant`?

ekpyron replied:
I'm not sure - for the abi it holds specifically that ``constant`` = (``pure`` or ``view``), so if people miss the ``constant`` field, that's what they should use instead - the fact that ``view`` replaces ``constant`` should rather be in the docs for state mutability itself and I think it's there already - I'll check.

chriseth commented on libsolidity/analysis/ViewPureChecker.cpp:
```diff
@@ -116,31 +116,23 @@ class AssemblyViewPureChecker: public boost::static_visitor<void>
-		for (ContractDefinition const* c: source->filteredNodes<ContractDefinition>(source->nodes()))
-			contracts[c] = enforceView;
+		auto const& sourceContracts = source->filteredNodes<ContractDefinition>(source->nodes());
+		contracts.insert(contracts.end(), sourceContracts.begin(), sourceContracts.end());
```
Can you use `contracts += source->filteredNodes<ContractDefinition>(source->nodes())`?

ekpyron replied:
Didn't know that that is defined for std::vector in libdevcore - I'll change it.

chriseth commented on test/contracts/LLL_ERC20.cpp:
```diff
@@ -216,39 +216,39 @@ static char const* erc20Code = R"DELIMITER(
 
       ;; ----------------------------------------------------------------------
       ;; Getter for the name of the token.
-      ;; @abi name() constant returns (string)
```
This is lll, do we really want to change it?

chriseth commented on test/libsolidity/ASTLegacyJSON.cpp:
```diff
@@ -266,15 +266,13 @@ BOOST_AUTO_TEST_CASE(function_type)
 	BOOST_CHECK_EQUAL(argument["attributes"]["name"], "x");
 	BOOST_CHECK_EQUAL(argument["attributes"]["type"], "function () payable external returns (uint256)");
 	Json::Value funType = argument["children"][0];
-	BOOST_CHECK_EQUAL(funType["attributes"]["constant"], false);
```
Can you add the corresponding check about stateMutability?

ekpyron replied:
Done.

ekpyron commented:
@chriseth Should we maybe disable the compilation tests on 050 already?

ekpyron commented:
I think I now caught all errors in the in-tree compilation tests. Not sure what to do about the external tests, though.

axic commented:
Changed the base and rebased.

axic commented on docs/abi-spec.rst:
```diff
@@ -313,7 +313,10 @@ A function description is a JSON object with the fields:
+
+.. warning::
+    Prior to version 0.5.0 there was an additional field ``constant``. In 0.5.0 this field has been removed.
+    Instead you can check whether stateMutability is ``pure`` or ``view``.
```
`stateMutability` should also be marked as a keyword (double backticks)

axic commented:
Why are tests failing?

Do you want to merge the commit of removing `constant` from the local test quite before there's a need for too much rebasing?

ekpyron commented:
@axic I think the test failure was due to some random RPC failure unrelated to this PR - I just rebased, so we'll see whether the tests fail again.
Otherwise I think it should be fine to merge it all at once, but I'll recheck, when I'm back in Berlin next week.

ekpyron commented:
The remaining test failures should be due to the external compilation tests. I think we should merge #4206 - I created a PR with the required changes to the gnosis contracts here: https://github.com/axic/pm-contracts/pull/1 ; zeppelin seems to have removed ``constant`` already.

axic commented on libsolidity/analysis/ViewPureChecker.cpp:
```diff
@@ -116,31 +116,23 @@ class AssemblyViewPureChecker: public boost::static_visitor<void>
-		bool enforceView = source->annotation().experimentalFeatures.count(ExperimentalFeature::V050);
-		for (ContractDefinition const* c: source->filteredNodes<ContractDefinition>(source->nodes()))
-			contracts.emplace_back(c, enforceView);
+		auto const& sourceContracts = source->filteredNodes<ContractDefinition>(source->nodes());
```
Does this needs this local variable or otherwise the overloaded `+=` operator won't see it as a `const&`?

ekpyron replied:
Actually, there's no reason for the local variable - I'll remove it.

axic commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -322,11 +322,15 @@ StateMutability Parser::parseStateMutability(Token::Value _token)
 		stateMutability = StateMutability::Pure;
+	else if (_token == Token::Constant)
+		fatalParserError(
+			"The state mutability modifier \"constant\" was removed in 0.5.0. "
```
Nice! I guess it makes sense to keep this at least for a while.

axic commented on test/libsolidity/ASTLegacyJSON.cpp:
```diff
@@ -266,15 +266,15 @@ BOOST_AUTO_TEST_CASE(function_type)
 	Json::Value funType = argument["children"][0];
-	BOOST_CHECK_EQUAL(funType["attributes"]["constant"], false);
+	BOOST_CHECK_EQUAL(funType["attributes"]["stateMutability"], "payable");
 	BOOST_CHECK_EQUAL(funType["attributes"]["payable"], true);
```
@chriseth @ekpyron Hm, do we also want to remove this and prefer `stateMutability`? Not as part of this PR though.

ekpyron replied:
Yes, I think it makes sense to remove ``payable`` as well and to always use ``stateMutability`` to avoid information duplication.

axic commented on test/libsolidity/SolidityABIJSON.cpp:
```diff
@@ -75,7 +75,6 @@ BOOST_AUTO_TEST_CASE(basic_test)
 	{
 		"name": "f",
-		"constant": false,
 		"payable" : false,
```
Same question applies here.

axic commented on test/compilationTests/zeppelin/crowdsale/Crowdsale.sol:
```diff
@@ -92,7 +92,7 @@ contract Crowdsale {
+  function validPurchase() internal returns (bool) {
     uint256 current = block.number;
     bool withinPeriod = current >= startBlock && current <= endBlock;
     bool nonZeroPurchase = msg.value != 0;
```
This is non-view because of `msg.value` IIRC, but why doesn't it require `payable` then?

This might be something we need to think about a bit.

axic commented on test/compilationTests/milestonetracker/RLP.sol:
```diff
@@ -376,7 +376,7 @@ library RLP {
 
  // Assumes that enough memory has been allocated to store in target.
- function _copyToBytes(uint btsPtr, bytes memory tgt, uint btsLen) private view {
+ function _copyToBytes(uint btsPtr, bytes memory tgt, uint btsLen) private {
```
Why is is non-view or even non-pure? It only does mstore/mload.

ekpyron replied:
As far as I remember it's the jump in the inline assembly that's not considered pure or even view at the moment (making sure that it always jumps to a piece of code that's view or pure is probably non-trivial).

axic replied:
Right, that makes sense, we should document this properly and how to avoid it: use if/for/switch.

axic commented on test/compilationTests/zeppelin/token/LimitedTransferToken.sol:
```diff
@@ -51,7 +51,7 @@ contract LimitedTransferToken is ERC20 {
    * specific logic for limiting token transferability for a holder over time.
    */
-  function transferableTokens(address holder, uint64 time) view public returns (uint256) {
+  function transferableTokens(address holder, uint64 time) public returns (uint256) {
```
Why is this non-view? `balanceOf` is still marked as `view`.

axic replied:
Hm, is it because the `interface` is declared non-view and this still doesn't let a stricter set to be set than in the interface?

ekpyron replied:
Actually, at the moment I don't see a reason for this one and the same one in ``VestedToken.sol`` not both to be view - I'll change it and see what the tests say.

ekpyron commented:
We also need to decide how to deal with ``DAO`` and ``truffle`` tests (unless I have missed a decision there).

axic commented:
Truffle has been fixed for all these things, but the projects using them aren't using the latest release yet. I've asked Gnosis, will check Zeppelin too.

As for the solc-js I'd prefer to create a DAO050 directory and keep the 0.4.x version intact because I'd like to run solc-js tests on different version as it should be the backwards compatibility glue.

axic commented:
Merged the solc-js changes and updated the `solcjsTests.sh` to use it in #4280. I think that should work, be merged, and then this PR should be close to a mergeable state.

axic commented:
Rebased.

axic commented:
Only the two external projects (gnosis, zeppelin) are failing. Getting close.

ekpyron commented:
I'm not entirely sure what to make of the 20000 line output of the still failing external tests...

chriseth commented:
Requires rebase.

axic commented:
Do we want to split out the (syntax) test updates?

ekpyron commented:
I split this up further and extracted #4409 and #4413. What remains for this PR is the JSON ABI change, but that still breaks the openzeppelin test suite.

ekpyron commented:
@axic @chriseth As far as I can tell, the external test failure is due to truffle relying on the ``constant`` field to be part of the json ABI. Not sure how to proceed (shall I look into modifying truffle? shall we open an issue at their repo? shall we keep ``constant`` in the ABI for the time being?).

ekpyron commented:
There's an issue for truffle https://github.com/trufflesuite/truffle/issues/1093
and a PR for web3  https://github.com/ethereum/web3.js/pull/1765 now.

ekpyron commented:
I just discussed with @chriseth that it may make sense to postpone this change for now - unless everything that depends on ``constant`` in the ABI is updated, it's likely to break a lot of existing javascript code - and only has very limited gain. We should still try to get external code updated, so that we may change this in the future.

axic commented on docs/abi-spec.rst:
```diff
@@ -412,7 +412,10 @@ A function description is a JSON object with the fields:
+
+.. warning::
+    Prior to version 0.5.0 there was an additional field ``constant``. In 0.5.0 this field has been removed.
+    Instead you can check whether ``stateMutability`` is ``pure`` or ``view``.
```
We should create a PR adding a deprecation warning for both `constant` and `payable`.

axic replied:
Did that in #4549.

axic commented:
This PR changes both the AST JSON and the ABI output. I think we can merge the AST JSON change, but as agreed the ABI should remain a little bit longer.

axic commented:
We should also submit a PR to go-ethereum's abigen and potentially parity's abigen.

Truffle 5.x was updated and they plan to backport the changes 4.x. @frozeman made web3 releases.

@yann300 @iurimatias is remix / embark making use of `stateMutability`?

axic commented:
@ekpyron can you create a PR with the AST changes only?

ekpyron commented:
@axic I will.

axic commented on docs/abi-spec.rst:
```diff
@@ -411,10 +411,14 @@ A function description is a JSON object with the fields:
 - ``constant``: ``true`` if function is either ``pure`` or ``view``, ``false`` otherwise.
 
-``type`` can be omitted, defaulting to ``"function"``, likewise ``payable`` and ``constant`` can be omitted, both defaulting to ``false``.
+.. warning::
```
See  the `warning` below, which has the same content.

iurimatias commented:
@axic we do and it's a breaking change for us, we'll need to update

axic commented:
@chriseth @ekpyron I think we can ~~close this now~~ postpone this now. Since I consider this a "protocol change", it doesn't necessarily needs to be aligned with breaking releases. Just the time has to be right for the ecosystem to adapt.

ekpyron commented:
@axic Yes, I think we were pretty much heading towards postponing this anyways and classifying this as ``protocol change`` and decoupling those from breaking releases makes sense.

chriseth commented:
Removed from 0.5.0.

axic reviewed (dismissed):
Needs to be rebased.

axic commented:
It is missing a changelog entry and all the tests are failing.

ekpyron commented:
Easy to recreate and unclear when we'll actually do this, so I'm closing this PR for now and added the issue to the backlog.
